test/archethic/contracts/interpreter/scope_test.exs

```diff
@@ -5,55 +5,589 @@ defmodule Archethic.Contracts.Interpreter.ScopeTest do
+             } = Process.get(:scope)
+    end
+
+    test "should correctly write in the current scope the current context" do
+      # Setup with multiple contexts
+      scope = %{
+        "context_list" => ["current_context", "other_context"],
+        "current_context" => %{
+          "scope_hierarchy" => ["my_scope"],
+          "my_scope" => %{}
+        },
+        "other_context" => %{
+          "scope_hierarchy" => []
+        }
+      }
+
+      Process.put(:scope, scope)
+
+      Scope.write_at("my_var", 42)
+
+      assert %{
+               "context_list" => ["current_context", "other_context"],
+               "current_context" => %{
+                 "scope_hierarchy" => ["my_scope"],
+                 "my_scope" => %{
+                   "my_var" => 42
+                 }
+               },
+               "other_context" => %{
+                 "scope_hierarchy" => []
+               }
+             } = Process.get(:scope)
+    end
   end
 
-  test "update_global/2" do
-    Scope.init(%{"transaction" => %{"content" => "cat"}})
-    Scope.update_global(["transaction"], fn t -> %{t | "content" => "dog"} end)
-    assert Scope.read_global(["transaction", "content"]) == "dog"
+  describe "write_cascade/2" do
+    test "should write the variable in the current scope if it doesn't exist anywhere" do
+      scope = %{
+        "context_list" => ["current_context"],
+        "current_context" => %{
+          "scope_hierarchy" => ["only_scope"],
+          "only_scope" => %{}
+        }
+      }
+
+      Process.put(:scope, scope)
+
+      Scope.write_cascade("my_var", 42)
+
+      assert %{
+               "context_list" => ["current_context"],
+               "current_context" => %{
+                 "scope_hierarchy" => ["only_scope"],
+                 "only_scope" => %{
+                   "my_var" => 42
+                 }
+               }
+             } = Process.get(:scope)
+    end
+
+    test "should overwrite the variable in the closest parent scope" do
+      scope = %{
+        "context_list" => ["current_context"],
+        "current_context" => %{
+          "scope_hierarchy" => ["first_scope", "second_scope"],
+          "first_scope" => %{
+            "my_var" => 24,
+            "second_scope" => %{
+              "another_var" => 99
+            }
+          }
+        }
+      }
+
+      Process.put(:scope, scope)
+
+      Scope.write_cascade("my_var", 42)
+
+      assert %{
+               "context_list" => ["current_context"],
+               "current_context" => %{
+                 "scope_hierarchy" => ["first_scope", "second_scope"],
+                 "first_scope" => %{
+                   "my_var" => 42,
+                   "second_scope" => %{
+                     "another_var" => 99
+                   }
+                 }
+               }
+             } = Process.get(:scope)
+    end
+
+    test "should correctly overwrite variable in closest parent scope among multiple parent scopes" do
```

Comment:
Normally this behavior should never happen since the second "my_var" => 24 should have overwrite the first "my_var" => 11

Reply:
actually does happen in for loops